pil-analyzer/src/type_inference.rs

```diff
@@ -638,17 +652,78 @@ impl<'a> TypeChecker<'a> {
                             self.expect_type_of_pattern(&value_type, pattern)?;
                         }
                         StatementInsideBlock::Expression(expr) => {
-                            self.expect_type_with_flexibility(self.statement_type, expr)?;
+                            self.check_kind_in_expression(expr)?
```

Comment:
I think this should just be
```suggestion
                            self.expect_type_with_flexibility(self.statement_type(), expr)?;
```
where the `statement_type` function returns the correct expected type depending on the current context / function kind.